protocolupgrade/engine.go

```diff
@@ -0,0 +1,252 @@
+	events             map[string]*eventspb.ProtocolUpgradeEvent
+	lock               sync.RWMutex
+
+	requiredMajority num.Decimal
+	upgradeStatus    *types.UpgradeStatus
+}
+
+func New(log *logging.Logger, config Config, broker Broker, topology ValidatorTopology) *Engine {
+	log = log.Named(namedLogger)
+	log.SetLevel(config.Level.Get())
+	return &Engine{
+		activeProposals: map[string]*protocolUpgradeProposal{},
+		events:          map[string]*eventspb.ProtocolUpgradeEvent{},
+		log:             log,
+		config:          config,
+		broker:          broker,
+		topology:        topology,
+		hashKeys:        []string{(&types.PayloadProtocolUpgradeProposals{}).Key()},
+		upgradeStatus:   &types.UpgradeStatus{},
+	}
+}
+
+func (e *Engine) OnRequiredMajorityChanged(_ context.Context, requiredMajority num.Decimal) error {
+	e.requiredMajority = requiredMajority
+	return nil
+}
+
+// UpgradeProposal records the intention of a validator to upgrade the protocol to a release tag at block height.
+func (e *Engine) UpgradeProposal(ctx context.Context, pk string, upgradeBlockHeight uint64, vegaReleaseTag, dataNodeReleaseTag string) error {
+	e.lock.RLock()
+	defer e.lock.RUnlock()
+	if !e.topology.IsTendermintValidator(pk) {
+		// not a tendermint validator, so we don't care about their intention
+		return nil
+	}
+
+	if upgradeBlockHeight <= e.currentBlockHeight {
+		return errors.New("upgrade block earlier than current block height")
+	}
+
+	if e.upgradeStatus.AcceptedReleaseInfo != nil {
+		return errors.New("protocol upgrade already scheduled")
+	}
+
+	if !e.topology.IsSelfTendermintValidator() {
+		return errors.New("validator is not a tendermint validator - cannot decide on upgrade")
+	}
+
+	_, err := semver.Parse(vegaReleaseTag)
+	if err != nil {
+		e.log.Error("invalid protocol version for upgrade received", logging.String("version", vegaReleaseTag), logging.Error(err))
+		return err
+	}
+
+	_, err = semver.Parse(dataNodeReleaseTag)
+	if err != nil {
+		e.log.Error("invalid protocol version for upgrade received", logging.String("version", vegaReleaseTag), logging.Error(err))
+		return err
+	}
+
+	ID := protocolUpgradeProposalID(upgradeBlockHeight, vegaReleaseTag, dataNodeReleaseTag)
+
+	// if it's the first time we see this ID, generate an active proposal entry
+	if _, ok := e.activeProposals[ID]; !ok {
+		e.activeProposals[ID] = &protocolUpgradeProposal{
+			blockHeight:        upgradeBlockHeight,
+			vegaReleaseTag:     vegaReleaseTag,
+			dataNodeReleaseTag: dataNodeReleaseTag,
+			accepted:           map[string]struct{}{},
+		}
+	}
+
+	active := e.activeProposals[ID]
+	active.accepted[pk] = struct{}{}
+	evt := events.NewProtocolUpgradeProposalEvent(ctx, active.blockHeight, active.vegaReleaseTag, active.dataNodeReleaseTag, active.approvers(), eventspb.ProtocolUpgradeProposalStatus_PROTOCOL_UPGRADE_PROPOSAL_STATUS_PENDING)
+	evtProto := evt.Proto()
+	e.events[ID] = &evtProto
+	e.broker.Send(evt)
+	return nil
```

Comment:
If a validator registered intent for another release before, should we remove them from the approvers of the release if they register for a new one?

Reply:
we can yeah